dawn/src/dawn/Validator/WeightChecker.cpp

```diff
@@ -0,0 +1,135 @@
+//===--------------------------------------------------------------------------------*- C++ -*-===//
+//                          _
+//                         | |
+//                       __| | __ ___      ___ ___
+//                      / _` |/ _` \ \ /\ / / '_  |
+//                     | (_| | (_| |\ V  V /| | | |
+//                      \__,_|\__,_| \_/\_/ |_| |_| - Compiler Toolchain
+//
+//
+//  This file is distributed under the MIT License (MIT).
+//  See LICENSE.txt for details.
+//
+//===------------------------------------------------------------------------------------------===//
+
+#include "WeightChecker.h"
+#include "dawn/AST/ASTFwd.h"
+#include "dawn/Support/SourceLocation.h"
+
+namespace dawn {
+
+void WeightChecker::WeightCheckerImpl::visit(
+    const std::shared_ptr<dawn::ast::FieldAccessExpr>& fieldAccessExpr) {
+  if(parentIsWeight_) {
+    auto fieldName = fieldAccessExpr->getName();
+    // the name in the FieldAccessExpr may be stale if the there are nested stencils
+    // in this case we need to look up the new AccessID in the data of the fieldAccessExpr
+    if(fieldAccessExpr->hasData()) {
+      auto newAccessID = fieldAccessExpr->getData<iir::IIRAccessExprData>().AccessID;
+      if(newAccessID.has_value()) {
+        DAWN_ASSERT(idToNameMap_.count(newAccessID.value()));
+        fieldName = idToNameMap_.at(newAccessID.value());
+      }
+    }
+
+    DAWN_ASSERT(nameToDimensions_.count(fieldName));
+    weightsValid_ = sir::dimension_cast<const sir::UnstructuredFieldDimension&>(
+                        nameToDimensions_.at(fieldName).getHorizontalFieldDimension())
+                        .isDense();
+  }
+}
+void WeightChecker::WeightCheckerImpl::visit(const std::shared_ptr<dawn::ast::FunCallExpr>& expr) {
+  if(parentIsWeight_) {
+    weightsValid_ = false;
+    return;
+  } else {
+    for(const auto& s : expr->getChildren()) {
+      s->accept(*this);
+    }
+  }
+}
+void WeightChecker::WeightCheckerImpl::visit(
+    const std::shared_ptr<dawn::ast::StencilFunCallExpr>& expr) {
+  if(parentIsWeight_) {
+    weightsValid_ = false;
+    return;
+  } else {
+    for(const auto& s : expr->getChildren()) {
```

Comment:
This will only visit the arguments, not the body of a function. I guess you are looking for a nested reduction here, so probably you should also forward the visit to the body.